### PR TITLE
Adds an upload service

### PR DESCRIPTION
**Pull request by atruskie**
This upload service allows for users to upload audio without having server ssh credentials.

This is part 1 of a larger set of changes. None of this functionality is exposed at the moment. Rather this change set just adds the service and makes sure it is usable.

The upload server itself is essentially a microservice. It is a go package that allows for sftp, webdav, and ftp uploading and has a REST based admin API.

We use the REST API to provision (and chroot) upload users that each harvest process will be able to use. Access is scoped, and temporary.

Note all the code in the lib/gems/sftpgo_generated_client directory was generated by the openapi code generation tool. The command used is stored in provision/sftpgo_schema_generation.sh

I also added `turnip` - a gherkin/cucumber implementation for rspec. Its BDD style makes it a litte easier to test complex services.

Also refactored Dockerfile with view to avoid deploy builds.

## Changes

Ignore lib/gems/sftpgo_generated_client

## Problems

The BawWorkers::UploadService::Communicator API is specualtive.

## Visual Changes

If the PR has any visual changes to the website, post pictures of the new pages and how they look.
Label any images if more than one is given.
image

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Remove/Reduce warnings from edited files
- [x] Unit tests have been added for changes
- [ ] Ensure CI build is passing

### CODE REVIEW

File: spec/lib/gems/baw_workers/upload_service_steps.rb [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
```
Comment (Allcharles) [inline spec/lib/gems/baw_workers/upload_service_steps.rb:1]: Some of these tests have no expectations, are they just testing if the function will crash or not?

**Review by Allcharles — APPROVED** [review]

File: lib/gems/sftpgo_generated_client/.travis.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+language: ruby
```
[the remark this answers is not in the document]
Comment (atruskie) [reply lib/gems/sftpgo_generated_client/.travis.yml:1]: Not worth customizing generated content - it just gets overwritten

File: provision/upload/config/httpd_auth [hunk 3]
```diff
@@ -0,0 +1 @@
+upload_admin:$apr1$M6w8zgYR$MwRdslSJXaB.zGPm7bW8/.
```
[the remark this answers is not in the document]
Comment (atruskie) [reply provision/upload/config/httpd_auth:1]: Yup, it just "password" hashed in the format sftpgo expects. Only used in dev setup

File: provision/upload/config/sftpgo.yaml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  #ciphers: []
+  #macs: []
+  #trusted_user_ca_keys: []
+  #login_banner_file: ''
+  enabled_ssh_commands:
+    - md5sum
+    - sha1sum
+    - cd
+    - pwd
+    - scp
```
[the remark this answers is not in the document]
Comment (atruskie) [reply provision/upload/config/sftpgo.yaml:34]: These are command enabled in addition to the standard SFTP protocol. These settings are the same as the default: https://github.com/drakkan/sftpgo/blob/bdf18fa862b952ed82f34cd94d3104e043050b99/sftpgo.json#L25-L30

File: spec/lib/gems/baw_workers/upload_service_steps.rb (hunk 1 above)
Comment (atruskie) [reply spec/lib/gems/baw_workers/upload_service_steps.rb:1]: Yeah, some are just setup steps. Some do tests.

**Pull request merged**